blog/2021-05-29-using-rust.md

```diff
@@ -0,0 +1,74 @@
+
+It's a big project with a lot of components - the AI models, the app, and the backend infrastructure which handles it all. We'll be taking a look at the backend in this article.
+
+## Lifecycle of an image
+
+Before we dive into the infrastructure, it would help to take a look at what an image goes through once it enters our system.
+
+- The image is assigned a UUID.
+- The image is uploaded to a `s3` bucket.
+- The image is then downloaded, preprocessed, run through our models, and a food tip is generated for the image.
+- The food tip, inference results and the image hash form the final response.
+
+## Stack decisions: Rust and haskell
+
+Haskell is used for the client facing API, and it's for other ad-hoc tasks such as reading/writing to a database among others.
+
+Rust is used for image preprocessing, model inference and sending the results back. Sending of results was done differently in the two approaches outlined below. Rust was chosen for it's high efficiency, small executable footprint, and absence of a garbage collector. It also had strong type system, speed & relatively actively maintained Tensorflow (client) library.
+
+# Introduction and pitfalls of the existing architecture
+
+![](/images/v2-arch-diagram.png)
+The internal organization of the rust service in this architecture is outlined above.
+There were 3 main parts, all running concurrently on 3 separate tokio runtimes - namely polling sqs, preprocessing and inferring from the images, and cleanup tasks.
+
+The external processes related to this architecture are outlined below.
+![](/images/v2-external-arch-diagram.png)
+
+The main gripe we had was in the `S3` to `SQS` upload event notification. In our experiments, it was [very slow](https://github.com/Holmusk/aws_benchmarks/tree/master/s3_event_to_sqs), and we aim for the service to have a very low latency, with the goal being that every image that comes into the system should be scored/rated in under 1 second. Because of the way the system was designed, this meant that we'd need a pretty big makeover on the rust side, and some tweaks on the haskell side. This is also mentioned in the [AWS docs](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html), where they point that `Typically, event notifications are delivered in seconds but can sometimes take a minute or longer`.
+
+# The new architecture
+
+Like mentioned before, the main reason for redesigning the architecture was to avoid the `S3` to `SQS` upload event notification as speed was of high importance. In the process, we found out that we actually simplified it, by removing unnecessary moving parts.
+
+Internally, the rust service now also has a webserver. The client facing API proxies the requests it receives to the rust webservers via a load balancer. In this version, we completely eliminate the use of SQS.
```

Comment:
```suggestion
Internally, the Rust service now also has a web server. The client facing API (written in Haskell) proxies the HTTP requests it receives to the Rust server via a load balancer (AWS ELB). In this version of the architecture (v3), we completely eliminate the use of a queue(SQS).
```

Link to what a web server is -> https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_is_a_web_server

This is to avoid any ambiguity and give readers with little context, a better chance at understanding the post

Reply:
looks good! I won't be mentioning `v3` or any version numbers anywhere though, because that might lead to more confusion. good to make webservers more clear tho ⚡